src/allmydata/node.py

```diff
@@ -268,30 +286,47 @@ class _Config(object):
     class; names and funtionality have been kept the same while moving
     the code. It probably makes sense for several of these APIs to
     have better names.
-    """
 
-    def __init__(self, configparser, portnum_fname, basedir, config_fname):
-        """
-        :param configparser: a ConfigParser instance
+    :ivar ConfigParser config: The actual configuration values.
 
-        :param portnum_fname: filename to use for the port-number file
-           (a relative path inside basedir)
+    :ivar str portnum_fname: filename to use for the port-number file (a
+        relative path inside basedir).
 
-        :param basedir: path to our "node directory", inside which all
-           configuration is managed
+    :ivar str _basedir: path to our "node directory", inside which all
+        configuration is managed.
 
-        :param config_fname: the pathname actually used to create the
-            configparser (might be 'fake' if using in-memory data)
-        """
-        self.portnum_fname = portnum_fname
-        self._basedir = abspath_expanduser_unicode(ensure_text(basedir))
-        self._config_fname = config_fname
-        self.config = configparser
-        self.nickname = self.get_config("node", "nickname", u"<unspecified>")
-        assert isinstance(self.nickname, str)
+    :ivar (FilePath|NoneType) config_path: The path actually used to create
+        the configparser (might be ``None`` if using in-memory data).
 
-    def validate(self, valid_config_sections):
-        configutil.validate_config(self._config_fname, self.config, valid_config_sections)
+    :ivar ValidConfiguration valid_config_sections: The validator for the
+        values in this configuration.
+    """
+    config = attr.ib(validator=attr.validators.instance_of(configparser.ConfigParser))
+    portnum_fname = attr.ib()
+    _basedir = attr.ib(
+        converter=lambda basedir: abspath_expanduser_unicode(ensure_text(basedir)),
+    )
+    config_path = attr.ib(
+        validator=attr.validators.optional(
+            attr.validators.instance_of(FilePath),
+        ),
+    )
+    valid_config_sections = attr.ib(
+        default=configutil.ValidConfiguration.everything(),
+        validator=attr.validators.instance_of(configutil.ValidConfiguration),
+    )
+
+    @property
+    def nickname(self):
+        nickname = self.get_config("node", "nickname", u"<unspecified>")
+        assert isinstance(nickname, str)
+        return nickname
+
+    @property
+    def _config_fname(self):
+        if self.config_path is None:
```

Comment:
Encouraging use of the FilePath is part of it but the other part is just preserving the status quo - the class had a `_config_fname` attribute before.  It was the string path and there was no FilePath version.  So this property is basically just backwards compat covering up the change to storing the FilePath instead of the string.  I didn't want to have to go find all the users of `_config_fname` and update them as part of this change.